### PR TITLE
Fix panic canonical.go

### PR DESCRIPTION
**Pull request by dimmaq**
Panic code

```go
package main

import (
	"log"
	"os"

	dkim "github.com/emersion/go-dkim"
)

func main() {
	r, _ := os.Open(`mail2.eml`)
	verifications, err := dkim.Verify(r)
	if err != nil {
		log.Fatal(err)
	}
	for _, v := range verifications {
		if v.Err == nil {
			log.Println("Valid signature for:", v.Domain)
		} else {
			log.Println("Invalid signature for:", v.Domain, v.Err)
		}
	}
}
```

```
c:\Users\Acer\go\src\test01>test01.exe
panic: runtime error: slice bounds out of range

goroutine 1 [running]:
bufio.(*Reader).fill(0xc04208a000)
        C:/Go/src/bufio/bufio.go:86 +0x226
bufio.(*Reader).WriteTo(0xc04208a000, 0x974098, 0xc042074190, 0x9740c0, 0xc04208a000, 0x3e7010b0001)
        C:/Go/src/bufio/bufio.go:491 +0x114
io.copyBuffer(0x974098, 0xc042074190, 0x5fe500, 0xc04208a000, 0x0, 0x0, 0x0, 0x53d040, 0xc042074100, 0x974098)
        C:/Go/src/io/io.go:382 +0x311
io.Copy(0x974098, 0xc042074190, 0x5fe500, 0xc04208a000, 0xc042074190, 0x0, 0x0)
        C:/Go/src/io/io.go:362 +0x6f
github.com/emersion/go-dkim.verify(0xc0420aa000, 0x1d, 0x20, 0x5fe500, 0xc04208a000, 0xc0420a2000, 0x19c, 0xc0420a2010, 0x18a, 0x488b98, ...)
        C:/Users/Acer/go/src/github.com/emersion/go-dkim/verify.go:293 +0xd57
github.com/emersion/go-dkim.Verify(0x5fe6c0, 0xc042066018, 0xc042066018, 0x0, 0x0, 0x5fd800, 0xc042063f08)
        C:/Users/Acer/go/src/github.com/emersion/go-dkim/verify.go:117 +0x46c
main.main()
        C:/Users/Acer/go/src/test01/main.go:12 +0x69
```

### CODE REVIEW

**Comment by dimmaq** [comment]
mail2.eml https://gist.github.com/dimmaq/b75f7c4e5bd3e198c71d34582fce44a2

File: canonical.go [hunk 1]
```diff
@@ -117,12 +117,8 @@ func (c *relaxedBodyCanonicalizer) Write(b []byte) (int, error) {
 			if len(c.crlfBuf) > 0 {
 				canonical = append(canonical, c.crlfBuf...)
 				c.crlfBuf = nil
-
-				if len(c.wspBuf) > 0 {
-					canonical = append(canonical, c.wspBuf...)
-					c.wspBuf = nil
-				}
```
Comment (emersion) [inline canonical.go:124]: Why is this part removed? We don't always want to add only a single space.

File: canonical.go [hunk 2]
```diff
@@ -135,7 +131,10 @@ func (c *relaxedBodyCanonicalizer) Write(b []byte) (int, error) {
 		c.written = true
 	}
 
-	return c.w.Write(canonical)
+	if _, err := c.w.Write(canonical); err != nil {
+		return len(b), err
+	}
+	return len(b), nil
```
Comment (emersion) [inline canonical.go:137]: You're right, `Write` should return the number of bytes written from `b`. However, you can simplify this to:

```go
_, err := c.w.Write(canonical)
return len(b), err
```

**Review by emersion — CHANGES_REQUESTED** [review]
Thanks for contributing to this library!

I've requested a few changes. Please make sure the tests pass!

File: canonical.go (hunk 1 above)
Comment (dimmaq) [reply canonical.go:124]: https://tools.ietf.org/html/rfc6376#section-3.4 
3.4.4.  The "relaxed" Body Canonicalization Algorithm
 Reduce all sequences of WSP within a line to a single SP
          character.

File: canonical.go (hunk 2 above)
Comment (dimmaq) [reply canonical.go:137]: yes, it's better and easier

File: canonical.go (hunk 1 above)
Comment (emersion) [reply canonical.go:124]: Okay, I was confused by the term "within a line", assuming at the beginning of a line was not within a line.

Please fix the test then!

**Comment by emersion** [comment]
Superseded by #4
